goose.go

```diff
@@ -1051,136 +1064,142 @@ func (c *cursor) Remainder() []ast.Stmt {
 	return s
 }
 
-func endsWithReturn(s ast.Stmt) bool {
+func (ctx Ctx) endsWithReturn(s ast.Stmt) bool {
 	if s == nil {
 		return false
 	}
 	switch s := s.(type) {
 	case *ast.BlockStmt:
-		return stmtsEndWithReturn(s.List)
+		return ctx.stmtsEndWithReturn(s.List)
 	default:
-		return stmtsEndWithReturn([]ast.Stmt{s})
+		return ctx.stmtsEndWithReturn([]ast.Stmt{s})
 	}
 }
 
-func endIncludesIf(s ast.Stmt) bool {
-	if s == nil {
-		return false
-	}
-	if s, ok := s.(*ast.BlockStmt); ok {
-		return stmtsEndIncludesIf(s.List)
-	}
-	return stmtsEndIncludesIf([]ast.Stmt{s})
-}
-
-func stmtsEndWithReturn(ss []ast.Stmt) bool {
+func (ctx Ctx) stmtsEndWithReturn(ss []ast.Stmt) bool {
 	if len(ss) == 0 {
 		return false
 	}
-	// TODO: should also catch implicit continue
-	switch ss[len(ss)-1].(type) {
+	switch s := ss[len(ss)-1].(type) {
 	case *ast.ReturnStmt, *ast.BranchStmt:
 		return true
-	}
-	return false
-}
-
-func stmtsEndIncludesIf(ss []ast.Stmt) bool {
-	if len(ss) <= 1 {
-		return false
-	}
-	if _, ok := ss[len(ss)-1].(*ast.IfStmt); ok {
-		return false
-	}
-	for _, item := range ss {
-		if _, ok := item.(*ast.IfStmt); ok {
-			return true
+	case *ast.IfStmt:
+		left := ctx.endsWithReturn(s.Body)
+		right := ctx.endsWithReturn(s.Else)
+		if left != right {
+			ctx.unsupported(s, "nested if statement branches differ on whether they return")
 		}
+		return left // or right, doesn't matter
 	}
 	return false
 }
 
-func (ctx Ctx) stmts(ss []ast.Stmt, loopVar *string) coq.BlockExpr {
+func (ctx Ctx) stmts(ss []ast.Stmt, usage ExprValUsage) coq.BlockExpr {
 	c := &cursor{ss}
 	var bindings []coq.Binding
+	var finalized bool
 	for c.HasNext() {
-		bindings = append(bindings, ctx.stmt(c.Next(), c, loopVar))
+		s := c.Next()
+		// ifStmt is special, it gets a chance to "wrap" the entire remainder
+		// to better support early returns.
+		switch s := s.(type) {
+		case *ast.IfStmt:
+			bindings = append(bindings, ctx.ifStmt(s, c.Remainder(), usage))
+			finalized = true
+			break // This would happen anyway since we consumed the iterator via "Remainder"
+		default:
+			if c.HasNext() {
+				// "finalized" will always be true.
+				binding, _ := ctx.stmtInBlock(s, ExprValLocal)
+				bindings = append(bindings, binding)
+			} else {
+				// The last statement is special: we propagate the usage and store "finalized"
+				binding, fin := ctx.stmtInBlock(s, usage)
+				bindings = append(bindings, binding)
+				finalized = fin
+			}
+		}
```

Comment:
Is there a nicer way to do this entire "finalized" thing? Basically for the last statement we want to know if it "used up" its control effect or not.